kadai1/task4233/eimg/eimg.go

```diff
@@ -0,0 +1,214 @@
+Package eimg encodes image files.
+Both before and after extension can be specified.
+Default setting is `.jpg/jpeg` -> `.png`
+
+e.g.)
+eimg ./img -f=gif -t=jpg
+
+*/
+package eimg
+
+import (
+	"flag"
+	"fmt"
+	"image"
+	"image/gif"
+	"image/jpeg"
+	"image/png"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+const (
+	version = "0.0.1"
+	msg     = "eimg v" + version + ", converts file extension\n"
+)
+
+var (
+	fr = flag.String("f", "jpeg", "file extension before executing")
+	to = flag.String("t", "png", "file extension after executing")
+)
+
+// Eimg struct has RootDir, FromDir, ToExt.
+// They are used to execute this commnad.
+type Eimg struct {
+	RootDir string
+	FromExt string
+	ToExt   string
+}
+
+// New makes Eimg instance.
+// Call this function if you use this package.
+func New() *Eimg {
+	return &Eimg{
+		RootDir: ".",
+		FromExt: "jpeg",
+		ToExt:   "png",
+	}
+}
+
+// Run converts file extension(from -> to).
+func (eimg *Eimg) Run() error {
+	if err := eimg.SetParameters(); err != nil {
+		return err
+	}
+
+	if err := eimg.ConvertExtension(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetParameters sets parameters for execution.
+func (eimg *Eimg) SetParameters() error {
+	if len(os.Args) == 1 {
+		fmt.Fprintf(os.Stderr, "Usage:\n\t./eimg [-f=fromExtension -t=toExtension] rootDir\n")
+		os.Exit(1)
+	}
+
+	// Init Parameters for tests like
+	// https://github.com/task4233/gopherdojo-studyroom/blob/kadai1-task4233/kadai1/task4233/eimg/eimg_test.go#L28-L29 .
+	*fr = "jpeg"
+	*to = "png"
+
+	// parse information
+	flag.Parse()
+	args := flag.Args()
+
+	// set information.
+	// validation implements in EncodeFile() for functional extension.
+	if fr != nil && *fr != "jpeg" {
+		eimg.FromExt = *fr
+	}
+	if to != nil && *to != "png" {
+		eimg.ToExt = *to
+	} else {
+		eimg.ToExt = "png"
+	}
+	if args[len(args)-1] != "." {
+		if _, err := os.Stat(args[len(args)-1]); err != nil {
+			return ErrInvalidPath.WithDebug(err.Error())
+		}
+		eimg.RootDir = args[len(args)-1]
+	}
+
+	return nil
+}
+
+// ConvertExtension converts extension by using set parameters.
+func (eimg *Eimg) ConvertExtension() error {
+	filePaths, err := eimg.GetFilePathsRec(eimg.RootDir)
+	if err != nil {
+		return ErrInvalidPath.WithDebug(err.Error())
+	}
+
+	// encode each file
+	for _, filePath := range filePaths {
+		extension := filepath.Ext(filePath)
+		if extension == "" {
+			continue
+		}
+
+		// filepath.Ext starts with "."
+		// e.g.) filepath.Ext(filePath) => .txt
+		if extension[1:] == eimg.FromExt {
+			err := eimg.EncodeFile(filePath)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// EncodeFile encodes file predefined arguments
+func (eimg *Eimg) EncodeFile(filePath string) error {
+	file, err := os.Open(filepath.Clean(filePath))
+	if err != nil {
+		return ErrInvalidPath.WithDebug(err.Error())
+	}
+	// ref: https://www.yunabe.jp/docs/golang_pitfall.html
+	defer func() {
+		cerr := file.Close()
+		if cerr == nil {
+			return
```

Comment:
これだとエラーが発生しても何もしない気がします...
```suggestion
		if cerr != nil {
			err = fmt.Errorf("%v: %w", cerr, err)
```

Reply:
ご指摘通りreturn前に何らかのメッセージを落とさないと、失敗理由が分からなくなるため、Stderrに出力するようにしました。